### PR TITLE
Add DevTools version to `flutter --version` and `flutter doctor -v` output.

### LINKED ISSUE #92029: `flutter --version` should print the DevTools version now that DevTools is shipping in the SDK

### PR DESCRIPTION
**Pull request by kenzieschmoll**
Fixes https://github.com/flutter/flutter/issues/92029.

New output:

`flutter --version`
```
Flutter 2.6.0-12.0.pre.597 • channel devtools-version • git@github.com:kenzieschmoll/flutter.git
Framework • revision f598cb2f35 (50 seconds ago) • 2021-11-04 10:33:02 -0700
Engine • revision 5140e30cec
Tools • Dart 2.15.0 (build 2.15.0-268.0.dev) • DevTools 2.7.0
```
`flutter doctor -v`
```
[✓] Flutter (Channel devtools-version, 2.6.0-12.0.pre.598, on macOS 11.6 20G165 darwin-x64, locale en)
    • Flutter version 2.6.0-12.0.pre.598 at /Users/kenzieschmoll/develop/flutter
    • Upstream repository git@github.com:kenzieschmoll/flutter.git
    • Framework revision ef43472850 (6 days ago), 2021-11-04 16:15:36 -0700
    • Engine revision 5140e30cec
    • Dart version 2.15.0 (build 2.15.0-268.0.dev)
    • DevTools version 2.7.0
...
```

### CODE REVIEW

File: packages/flutter_tools/lib/src/cache.dart [hunk 1]
```diff
@@ -356,6 +356,30 @@ class Cache {
     }
   }
 
+  String get devToolsVersion {
+    if (_devToolsVersion == null) {
+      final Directory devToolsDir = getCacheDir('dart-sdk/bin/resources/devtools');
+      if (!devToolsDir.existsSync()) {
+        throwToolExit('Could not determine DevTools version.');
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/cache.dart:363]: We're assuming this directory exists, I'm thinking that if the directory is missing this is bug on our part, and thus we should probably throw some other kind of exception that we see in crash logging?

File: packages/flutter_tools/lib/src/cache.dart [hunk 2]
```diff
@@ -356,6 +356,30 @@ class Cache {
...
+      }
+      final File versionFile = _fileSystem.file('${devToolsDir.path}/version.json');
+      if (!versionFile.existsSync()) {
+        throwToolExit('Could not determine DevTools version.');
+      }
+      final dynamic data = jsonDecode(versionFile.readAsStringSync());
+      if (data is! Map<String, Object>) {
+        throwToolExit('Could not determine DevTools version.');
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/cache.dart:371]: and this should be a different exception, making it clear we found the file but failed to parse it

File: packages/flutter_tools/lib/src/version.dart [hunk 3]
```diff
@@ -159,7 +161,7 @@ class FlutterVersion {
     final String flutterText = 'Flutter$versionText • channel $channel • ${repositoryUrl ?? 'unknown source'}';
     final String frameworkText = 'Framework • revision $frameworkRevisionShort ($frameworkAge) • $frameworkCommitDate';
     final String engineText = 'Engine • revision $engineRevisionShort';
-    final String toolsText = 'Tools • Dart $dartSdkVersion';
+    final String toolsText = 'Tools • Dart $dartSdkVersion • DevTools $devToolsVersion';
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/version.dart:164]: This LGTM. We should also add this to the doctor (which is unfortunately a different codepath): https://github.com/christopherfujino/flutter/blob/f598cb2f35c1de110f2a809429dbfb41bc8d8e84/packages/flutter_tools/lib/src/doctor.dart#L433

File: packages/flutter_tools/lib/src/cache.dart (hunk 1 above)
Comment (kenzieschmoll) [reply packages/flutter_tools/lib/src/cache.dart:363]: Is there a specific exception that will show up in crash logs? Or another place in flutter_tools that does this that I could mimc here?

Comment (christopherfujino) [reply packages/flutter_tools/lib/src/cache.dart:363]: Pretty much anything other than ToolExit will show up in crash logging. You could throw a bare `Exception`, as we do elsewhere in this file, I'm not picky :)

Comment (kenzieschmoll) [reply packages/flutter_tools/lib/src/cache.dart]: done.

File: packages/flutter_tools/lib/src/cache.dart (hunk 2 above)
Comment (kenzieschmoll) [reply packages/flutter_tools/lib/src/cache.dart]: done

File: packages/flutter_tools/lib/src/version.dart (hunk 3 above)
Comment (kenzieschmoll) [reply packages/flutter_tools/lib/src/version.dart:164]: done - ptal

**Comment by kenzieschmoll** [comment]
@christopherfujino is it safe to throw exceptions when finding the DevTools version fails? Or is there a way to log the exception to crash reporting but fail gracefully?

Example test failure here: https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8831452341122609393/+/u/run_test.dart_for_tool_tests_shard_and_subshard_general/test_stdout.

Is there ever a case where `dart-sdk/bin/resources/devtools` will not be present? Trying to understand if this failure is expected and we really should just be failing gracefully.

**Comment by christopherfujino** [comment]
> @christopherfujino is it safe to throw exceptions when finding the DevTools version fails? Or is there a way to log the exception to crash reporting but fail gracefully?
> 
> Example test failure here: https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8831452341122609393/+/u/run_test.dart_for_tool_tests_shard_and_subshard_general/test_stdout.
> 
> Is there ever a case where `/cache/dart-sdk/bin/resources/devtools` will not be present? Trying to understand if this failure is expected and we really should just be failing gracefully.

I don't think we should ever expect this to happen. I think it's fine to leave it as an exception in this (in answer to your other question, we don't really have support for sending a crash report and then failing gracefully).

**Comment by christopherfujino** [comment]
I pushed a fix to the general_test and merged back upstream which should fix the integration tests.

**Review by christopherfujino — APPROVED** [review]
LGTM

**Comment by aam** [comment]
this seems to have broken flutter hhh buildbot https://logs.chromium.org/logs/dart/buildbucket/cr-buildbucket/8830752932360531713/+/u/flutter_test_tool_tests/stdout

**Comment by kenzieschmoll** [comment]
Is that failure a dupe of https://github.com/flutter/flutter/issues/93553 CC @christopherfujino @zanderso @jason-simmons

**Comment by jason-simmons** [comment]
yes - if `test/general.shard/cache_test.dart` is failing then try rebasing to the current framework head

**Comment by aam** [comment]
aha, thanks!

**Pull request merged**